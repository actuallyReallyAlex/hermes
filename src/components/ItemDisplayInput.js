import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Box, Button } from 'grommet'
import { Add } from 'grommet-icons'
import { connect } from 'react-redux'
import { storeCargo } from '../redux/actions/ship'
import { removeCash } from '../redux/actions/user'
import { removeItem } from '../redux/actions/world'

const ItemDisplayInput = ({
  handleStoreCargo,
  item,
  shipCargoVolumeRemaining,
  userCash
}) => {
  const { id, quantity } = item

  const [value, setValue] = useState(0)

  return (
    <Box gap="small" pad="medium">
      <Button
        data-testid={`add-button-${id}`}
        disabled={
          shipCargoVolumeRemaining === 0 ||
          value === 0 ||
          userCash < value * item.value
        }
        hoverIndicator
        icon={<Add />}
        onClick={() => handleStoreCargo(item, value)}
        plain
      />
      <label htmlFor={`quantity-input-${id}`}>Quantity to Add</label>
      <input
        id={`quantity-input-${id}`}
        max={
          shipCargoVolumeRemaining > quantity
            ? quantity
            : shipCargoVolumeRemaining
        }
        min={0}
        onChange={e => setValue(Number(e.target.value))}
        type="number"
        value={value}
      />
    </Box>
  )
}

ItemDisplayInput.propTypes = {
  handleStoreCargo: PropTypes.func.isRequired,
  item: PropTypes.object.isRequired,
  shipCargoVolumeRemaining: PropTypes.number.isRequired,
  userCash: PropTypes.number.isRequired
}

const mapStateToProps = ({ ship, user }) => ({
  shipCargoVolumeRemaining: ship.cargo.volumeRemaining,
  userCash: user.cash
})

const mapDispatchToProps = dispatch => ({
  handleStoreCargo: (item, quantity) => {
    // * dispatch an action to buy the items with the user's cash
    dispatch(removeCash(item.value * quantity))
    // * dispatch an action to store the item in ship cargo
    dispatch(storeCargo(item, quantity))
    // * dispatch an action to remove the item from the list of stored items on this planet
    dispatch(removeItem(item, quantity))
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ItemDisplayInput)
