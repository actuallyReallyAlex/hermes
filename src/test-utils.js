import React from 'react'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { defaultState } from './fixtures'
import { render } from '@testing-library/react'

const mockThunks = ({ dispatch, getState }) => next => action => {
  const mockedAction = () => ({ type: 'MOCKED ACTION' })

  let returnValue

  if (typeof action === 'function' || action === undefined) {
    returnValue = mockedAction
  } else {
    returnValue = next(action)
  }

  return returnValue
}

const middlewares = [mockThunks]

export const mockStore = configureStore(middlewares)

export const customRender = options => {
  const { component, state, props } = options

  const Component = component

  return render(
    <Provider store={mockStore(state ? state : defaultState)}>
      <Component {...props} />
    </Provider>
  )
}
