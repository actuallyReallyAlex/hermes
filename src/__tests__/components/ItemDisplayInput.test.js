import { customRender } from '../../test-utils'
import { fireEvent } from '@testing-library/react'
import ItemDisplayInput from '../../components/ItemDisplayInput'

const item = {
  name: 'Test Item',
  space: 1,
  value: 1,
  id: '0',
  destination: {
    name: 'Test Planet 2',
    value: 50
  },
  quantity: 10
}

describe('<ItemDisplayInput />', () => {
  it('Should render the <ItemDisplayInput /> component.', () => {
    const container = customRender({
      component: ItemDisplayInput,
      props: { item }
    })
    expect(container.asFragment()).toMatchSnapshot()
  })

  it('Should handle the user buying an item.', () => {
    const container = customRender({
      component: ItemDisplayInput,
      props: { item }
    })

    const { getByTestId } = container

    const quantityInput = container.container.querySelector('#quantity-input-0')
    const addButton = getByTestId('add-button-0')

    fireEvent.change(quantityInput, { target: { value: 2 } })
    fireEvent.click(addButton)
  })
})
