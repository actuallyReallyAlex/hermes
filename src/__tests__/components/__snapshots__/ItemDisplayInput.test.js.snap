// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemDisplayInput /> Should render the <ItemDisplayInput /> component. 1`] = `
<DocumentFragment>
  <div
    class="StyledBox-sc-13pk1d4-0 hroMnY"
  >
    <button
      class="StyledButton-sc-323bzc-0 jJyttA"
      data-testid="add-button-0"
      disabled=""
      type="button"
    >
      <svg
        aria-label="Add"
        class="StyledIcon-ofa7kd-0 iOkQrb"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,22 L12,2 M2,12 L22,12"
          fill="none"
          stroke="#000"
          stroke-width="2"
        />
      </svg>
    </button>
    <div
      class="StyledBox__StyledBoxGap-sc-13pk1d4-1 hPsznL"
    />
    <label
      for="quantity-input-0"
    >
      Quantity to Add
    </label>
    <div
      class="StyledBox__StyledBoxGap-sc-13pk1d4-1 hPsznL"
    />
    <input
      id="quantity-input-0"
      max="5"
      min="0"
      type="number"
      value="0"
    />
  </div>
</DocumentFragment>
`;
